Implement a language agnostic hardware representation

RhysGretsch81 opened the pull request.

Commit:
Implement a language agnostic hardware representation

Commit:
Implement a language agnostic hardware representation

mdko commented:
Hi @RhysGretsch81!

Thanks for the PR. Some requests:

- Could you add a detailed description here in the conversation about the motivation for this change, what problem it solves, and an example or two?
- Could you also add some unit tests to `test_importexport.py` (CodeCov is showing a large decrease in overall coverage)?
- Does a new file (`hardwareSchema.py`) necessarily need to be created? Or would this all work by just instead putting the `hardwareSchema()` function in `importexport.py`?
- (Nitpick): Could you change the name of `hardwareSchema()` to `hardware_schema()` (for the function) and to`hardwareschema.py` (for the file, if it ends up needing to stay in, depending on the answer to the previous bullet point)? This makes it more consistent with other naming conventions currently used.

(Also, is this related at all to JSON formats produced by things like Yosys?)

RhysGretsch81 commented:
> Hi @RhysGretsch81!
> 
> Thanks for the PR. Some requests:
> 
> * Could you add a detailed description here in the conversation about the motivation for this change, what problem it solves, and an example or two?
> * Could you also add some unit tests to `test_importexport.py` (CodeCov is showing a large decrease in overall coverage)?
> * Does a new file (`hardwareSchema.py`) necessarily need to be created? Or would this all work by just instead putting the `hardwareSchema()` function in `importexport.py`?
> * (Nitpick): Could you change the name of `hardwareSchema()` to `hardware_schema()` (for the function) and to`hardwareschema.py` (for the file, if it ends up needing to stay in, depending on the answer to the previous bullet point)? This makes it more consistent with other naming conventions currently used.
> 
> (Also, is this related at all to JSON formats produced by things like Yosys?)

RhysGretsch81 commented:
> Hi @RhysGretsch81!
> 
> Thanks for the PR. Some requests:
> 
> * Could you add a detailed description here in the conversation about the motivation for this change, what problem it solves, and an example or two?
> * Could you also add some unit tests to `test_importexport.py` (CodeCov is showing a large decrease in overall coverage)?
> * Does a new file (`hardwareSchema.py`) necessarily need to be created? Or would this all work by just instead putting the `hardwareSchema()` function in `importexport.py`?
> * (Nitpick): Could you change the name of `hardwareSchema()` to `hardware_schema()` (for the function) and to`hardwareschema.py` (for the file, if it ends up needing to stay in, depending on the answer to the previous bullet point)? This makes it more consistent with other naming conventions currently used.
> 
> (Also, is this related at all to JSON formats produced by things like Yosys?)

Hi @mdko 

- This change was motivated by Dr. Sherwood.  He wanted a language agnostic hardware representation to allow for easier communication with other hardware languages, as well as allowing a simple hardware representation for analysis without having to deal with language processing.  For more information on the motivation I would ask him.
- Sure, I am working on adding tests now
- The hardwareSchema is a way to validate that the JSON fits the standard required for importing/exporting.  I believe its useful to keep this as a separate file so other applications can use this schema simply by grabbing this file instead of having to copy/paste.
- I can update the naming conventions.  I'll add a commit to this PR with the updated names and unit tests

I was unaware of Yosys' JSON representation.  I'll look into seeing if I should align the protocols

Commit:
Update protocol for better understanding

Commit:
Create test cases for json import/export

RhysGretsch81 committed:
Rename hardwareSchema.py to hardwareschema.py

Commit:
Expose json input/output functions to users

Commit:
Merge branch 'hwRep' of https://github.com/RhysGretsch81/PyRTL into hwRep

RhysGretsch81 commented:
I have updated the protocols based off of a conversation with Dr. Sherwood.
I have also added test cases for both the import and export functions, as well as updating the naming conventions
